engine/access/state_stream/streamer.go

```diff
@@ -0,0 +1,107 @@
+package state_stream
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/module/executiondatasync/execution_data"
+	"github.com/onflow/flow-go/storage"
+)
+
+// Streamable represents a subscription that can be streamed.
+type Streamable interface {
+	ID() string
+	Close()
+	Fail(error)
+	Send(context.Context, interface{}, time.Duration) error
+	Next(context.Context) (interface{}, error)
+}
+
+// Streamer
+type Streamer struct {
+	log         zerolog.Logger
+	broadcaster *engine.Broadcaster
+	sendTimeout time.Duration
+	sub         Streamable
+}
+
+func NewStreamer(
+	log zerolog.Logger,
+	broadcaster *engine.Broadcaster,
+	sendTimeout time.Duration,
+	sub Streamable,
+) *Streamer {
+	return &Streamer{
+		log:         log.With().Str("sub_id", sub.ID()).Logger(),
+		broadcaster: broadcaster,
+		sendTimeout: sendTimeout,
+		sub:         sub,
+	}
+}
+
+// Stream is a blocking method that streams data to the subscription until either the context is
+// cancelled or it encounters an error.
+func (s *Streamer) Stream(ctx context.Context) {
+	s.log.Debug().Msg("starting streaming")
+	defer s.log.Debug().Msg("finished streaming")
+
+	notifier := s.broadcaster.Subscribe()
+
+	// always check the first time. This ensures that streaming continues to work even if the
+	// execution sync is not functioning (e.g. on a past spork network, or during an temporary outage)
+	notifier.Notify()
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.sub.Fail(fmt.Errorf("client disconnected: %w", ctx.Err()))
+			return
+		case <-notifier.Channel():
+			s.log.Debug().Msg("received broadcast notification")
+		}
+
+		err := s.sendAllAvailable(ctx)
+
+		if err != nil {
+			s.sub.Fail(err)
+			return
+		}
+	}
+}
+
+// sendAllAvailable reads data from the streamable and sends it to the client until no more data is available.
+func (s *Streamer) sendAllAvailable(ctx context.Context) error {
```

Comment:
nit: this function is basically Next+Send but there is quite a bit of code in between.  Maybe it is possible to simplify this a bit by moving code around (e.g. moving some for the error-processing code inside `if err`)

Reply:
it's mostly logging code. I'll clean it up